aten/src/ATen/core/jit_type.h

```diff
@@ -1710,4 +1712,54 @@ struct CAFFE2_API InterfaceType : public NamedType {
   bool is_module_;
 };
 
+template <TypeKind K>
+struct EnumerationType : public Type {
```

Comment:
It is hard to review this PR without the other context surrounding it. For example, currently `EnumerationType` provides no value since it does basically nothing. I'm sure at some point there will be common methods that are useful to define for enumerable types, but at this point I don't know what they are.